src/logging.cpp

```diff
@@ -222,23 +226,43 @@ void BCLog::Logger::LogPrintStr(const std::string &str)
     if (m_print_to_file) {
         std::lock_guard<std::mutex> scoped_lock(m_file_mutex);
 
+        // While the system is initializing, save some reasonable
+        // number of log messages for re-logging when the log
+        // file rotates (gets reopened).
+        if (m_relog_enable && m_msgs_relog.size() < 1000) {
+            m_msgs_relog.push_back(str_prefixed);
+        }
+
         // buffer if we haven't opened the log yet
         if (m_fileout == nullptr) {
             m_msgs_before_open.push_back(str_prefixed);
         }
         else
         {
+            FileWriteStr(str_prefixed, m_fileout);
+
             // reopen the log file, if requested
-            if (m_reopen_file) {
+            if (m_reopen_file && m_started_new_line) {
```

Comment:
Adding this condition fixes a minor bug that could have been fixed separately. I noticed this bug while testing this PR, but it's not very likely to happen in normal conditions. If the client receives a `SIGHUP` during a sequence of `LogPrintf()` (or `LogPrint()`) calls that don't end in a newline (the line is built up in separate calls, as here: https://github.com/bitcoin/bitcoin/blob/f6f924811d5b4ae2a6c77105403747cd31d3e49d/src/validation.cpp#L2133-L2140), then the first part of the line will be in the older `debug.log` file and the second part in the newer file. No information is lost, but it's a little strange. This change ensures that lines are not split across files.